FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,87 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				if let cache = try ManagedCache.find(in: context) {
+					completion(.found(
+						feed: cache.feed
+							.compactMap { ($0 as? ManagedFeedImage) }
+							.map {
+								LocalFeedImage(id: $0.id,
+								               description: $0.imageDescription,
+								               location: $0.location,
+								               url: $0.url)
+							},
+						timestamp: cache.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		let context = self.context
+		context.perform {
+			do {
+				let managedCache = try ManagedCache.getNewUniqueInstance(context: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = NSOrderedSet(array: feed.map { local in
+					let managed = ManagedFeedImage(context: context)
+					managed.id = local.id
+					managed.imageDescription = local.description
+					managed.location = local.location
+					managed.url = local.url
+					return managed
+				})
+				try context.save()
+				completion(nil)
+			} catch {
+				context.reset()
```

Comment:
`context.reset` invalidates all the managed objects of this context which in turn can be an issue if some of these objects are in use. To only reset the changes `context.rollback` needs to be used. 

Please replace `reset` call with `rollback`

```suggestion
				context.rollback()
```